pkg/transfer/flex/main.go

```diff
@@ -128,91 +154,385 @@ func (volp *DatasetVolumes) readDatasetOpts(mountPath string) (*datasetOpts, err
-		if err == nil { //if there was no error during upload remove all data
-			err = os.RemoveAll(mountPath)
-			
-			if err == nil {
-				err = volp.deleteDatasetOpts(mountPath)
-			}
+		if err != nil {
+			volp.deleteDatasetOpts(volp.getPath(kubeMountPath, RelPathOptions))
 		}
 	}()
 
-	if dsopts.Output == nil {
-		return nil //no output dataset, do nothing with the volume data
+	if err != nil {
+		return errors.Wrap(err, "failed to write volume database")
 	}
 
-	var trans transfer.Transfer
-	if trans, err = transfer.NewS3(&transfer.S3Conf{
-		Bucket: dsopts.Output.Bucket,
-	}); err != nil {
-		return err
+	//Set up input
+	err = volp.provisionInput(volp.getPath(kubeMountPath, RelPathInput), dsopts.Input)
+
+	defer func() {
+		if err != nil {
+			volp.destroyInput(volp.getPath(kubeMountPath, RelPathInput))
+		}
+	}()
+
+	if err != nil {
+		return errors.Wrap(err, "failed to provision input")
 	}
 
-	ref := &transfer.Ref{
-		Bucket: dsopts.Output.Bucket,
-		Key:    dsopts.Output.Key,
+	//Create volume to contain pod writes
+	err = volp.createFSInFile(volp.getPath(kubeMountPath, RelPathFSInFile), FileSystemExt4, WriteSpace)
+
+	defer func() {
+		if err != nil {
+			volp.destroyFSInFile(volp.getPath(kubeMountPath, RelPathFSInFile))
+		}
+	}()
+
+	if err != nil {
+		return errors.Wrap(err, "failed to create file system in a file")
 	}
 
-	_, err = trans.Upload(context.Background(), ref, mountPath)
+	//Mount the file system
+	err = volp.mountFSInFile(
+		volp.getPath(kubeMountPath, RelPathFSInFile),
+		volp.getPath(kubeMountPath, RelPathFSInFileMount),
+	)
+
+	defer func() {
+		if err != nil {
+			volp.unmountFSInFile(volp.getPath(kubeMountPath, RelPathFSInFileMount))
+		}
+	}()
+
 	if err != nil {
-		return err
+		return errors.Wrap(err, "failed to mount file system in a file")
+	}
+
+	//Set up overlay file system using input and writable fs-in-file
+	err = volp.mountOverlayFS(
+		filepath.Join(volp.getPath(kubeMountPath, RelPathFSInFileMount), "upper"),
+		filepath.Join(volp.getPath(kubeMountPath, RelPathFSInFileMount), "work"),
+		volp.getPath(kubeMountPath, RelPathInput),
+		kubeMountPath,
+	)
+
+	defer func() {
+		if err != nil {
+			volp.unmountOverlayFS(
+				filepath.Join(volp.getPath(kubeMountPath, RelPathFSInFileMount), "upper"),
+				filepath.Join(volp.getPath(kubeMountPath, RelPathFSInFileMount), "work"),
+				kubeMountPath,
+			)
+		}
+	}()
+
+	if err != nil {
+		return errors.Wrap(err, "failed to mount overlayfs")
+	}
+
+	return nil
+}
+
+//Unmount the flex volume
+func (volp *DatasetVolumes) Unmount(kubeMountPath string) (err error) {
+	// Upload any output
+	var dsopts *datasetOpts
+	dsopts, err = volp.readDatasetOpts(volp.getPath(kubeMountPath, RelPathOptions))
+	if err != nil {
+		return errors.Wrap(err, "failed to read volume database")
+	}
+
+	err = volp.handleOutput(kubeMountPath, dsopts.Output)
+	if err != nil {
+		return errors.Wrap(err, "failed to upload output")
+	}
+
+	//Clean up (as much as possible)
+	errs := []error{
```

Comment:
Usually, in go, we would not create a slice of errors but we would check each error one by one, and this would avoid to have an error nested in an error nested in another one. I'm not used to see this way of checking errors, so I would really like to get your opinion on that!

Reply:
I did it this way to do as much cleanup as possible, even if some of it fails. I took the advice from this post on StackOverflow:

https://stackoverflow.com/questions/15397419/go-handling-multiple-errors-elegantly

I asked @advanderveer for some advice and he suggested to use this package to handle and return the (possibly) multiple errors elegantly:

https://github.com/hashicorp/go-multierror

I've now changed the code to use that since we already had it as dependency anyway.